ykpersonalize.1
.\" Copyright (c) 2009-2014 Yubico AB
.\" Copyright (C) 2009, 2010 Tollef Fog Heen <tfheen@err.no>
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions are
.\" met:
.\"
.\"     * Redistributions of source code must retain the above copyright
.\"       notice, this list of conditions and the following disclaimer.
.\"
.\"     * Redistributions in binary form must reproduce the above
.\"       copyright notice, this list of conditions and the following
.\"       disclaimer in the documentation and/or other materials provided
.\"       with the distribution.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
.\" "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
.\" LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
.\" A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
.\" OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
.\" SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
.\" LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
.\" DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
.\" THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
.\" (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
.\" OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
.\"
.\" The following commands are required for all man pages.
.de URL
\\$2 \(laURL: \\$1 \(ra\\$3
..
.if \n[.g] .mso www.tmac
.TH ykpersonalize "1" "August 2009" "yubikey-personalization"
.SH NAME
ykpersonalize - personalize YubiKey OTP tokens
.SH SYNOPSIS
.B ykpersonalize
[\fI-1\fR | \fI-2\fR] [\fI-sfile\fR] [\fI-ifile\fR] [\fI-fformat\fR] [\fI-axxx\fR] [\fI-cxxx\fR]
[\fI-ooption\fR] [\fI-y\fR] [\fI-v\fR] [\fI-d\fR] [\fI-h\fR] [\fI-n\fR] [\fI-t\fR] [\fI-u\fR] [\fI-x\fR]
[\fI-z\fR] [\fI-m\fR] [\fI-S\fR] [\fI-V\fR]
.\".SH DESCRIPTION
.\" Add any additional description here
.SH OPTIONS
.PP
Set the AES key, user ID and other settings in a YubiKey.  For the complete
explanation of the meaning of all parameters, see the reference
manual:
.URL "http://yubico.com/files/YubiKey_manual-2.0.pdf" "YubiKey manual"
.TP
\fB\-1\fR
change the first configuration.  This is the default and is
normally used for true OTP generation.  In this configuration,
the option flag \fB-oappend-cr\fR is set by default.
.TP
\fB\-2\fR
change the second configuration.  This is for YubiKey II only and is
then normally used for static key generation.  In this configuration,
the option flags \fB-oappend-cr\fR, \fB-ostatic-ticket\fR, \fB-ostrong-pw1\fR,
\fB-ostrong-pw2\fR and \fB-oman-update\fR are set by default.
.TP
\fB-z\fR
delete configuration in selected slot
.TP
\fB\-s\fIfile\fR
save configuration to file instead of key.
(if file is -, send to stdout)
.TP
\fB\-i\fIfile\fR
read configuration from file.
(if file is -, read from stdin)
Configuration import is only valid for the ycfg format.
.TP
\fB\-f\fIformat\fR
format to be used with \fB-s\fR and \fB-i\fR.
Valid options are \fBycfg\fR and \fBlegacy\fR.
.TP
\fB\-a\fR[\fIxxx\fR]
the AES secret key as a 32 (or 40 for OATH-HOTP/HMAC CHAL-RESP) char hex value (not modhex) (none to prompt for key on stdin)
If \fB\-a\fR is not used a random key will be generated.
.TP
\fB\-c\fIxxx\fR
A 12 char hex value (not modhex) to use as access code for
programming.
NOTE: this does NOT SET the access code, that's done with \fB-oaccess\fI=\fR.
.TP
\fB\-o\fIoption\fR
change configuration option.  Possible option arguments are
.RS
.TP
\fBfixed\fR=\fIfffffffffff\fR
The modhex \fIpublic identity\fR of the YubiKey, 0-32 characters long (encoding
up to 16 bytes).
It's possible to give the identity in hex as well, just prepend the
value with `h:'. The fixed part is emitted before the OTP when the
button on the YubiKey is pressed. It can be used as an identifier for
the user, for example.
.TP
\fBuid\fR=\fIuuuuuu\fR
The uid part of the generated OTP, also called \fIprivate identity\fR, in hex.
Must be 12 characters long. The uid is 6 bytes of static data that is included
(encrypted) in every OTP, and is used to validate that an OTP was in fact encrypted
with the AES key shared between the YubiKey and the validation service. It cannot
be used to identify the YubiKey as it is only readable to those that know
the AES key.
.TP
\fBaccess\fR=\fIfffffffffff\fR
New hex access code to set. Must be 12 characters long.
If an access code is set, it will be required for subsequent reprogramming of the YubiKey.
.TP
\fBoath-imf\fR=\fIxxx\fR
Set OATH Initial Moving Factor. This is the initial counter value for the YubiKey.
This should be a value between 0 and 1048560, evenly dividable by 16.
.TP
[\-]\fIticket-flag\fR
Set/clear ticket flag, see the section `Ticket flags\&'
.TP
[\-]\fIconfiguration-flag\fR
Set/clear ticket flag, see the section `Configuration flags\&'
.RE
.TP
\fB-y\fR
always commit without prompting
.TP
\fB-d\fR
dry-run, run without writing a YubiKey
.TP
\fB-v\fR
Be more verbose
.TP
\fB-h\fR
Help
.TP
\fB-V\fR
Version
.TP
\fBYubiKey Neo only\fR
.RS
.TP
\fB-m mode\fR
set device configuration for the YubiKey.  It is parsed in the form
\fImode:cr_timeout:autoeject_timeout\fR
.br
where mode is:
.br
0    OTP device only.
.br
1    CCID device only.
.br
2    OTP/CCID composite device.
.br
3    U2F device only.
.br
4    OTP/U2F composite device.
.br
5    U2F/CCID composite device.
.br
6    OTP/U2F/CCID composite device.
.br
Add 80 to set MODE_FLAG_EJECT, for example: 81
.br
cr_timeout is the timeout in seconds for the YubiKey to wait on button press for challenge response (default is 15)
.br
autoeject_timeout is the timeout in seconds before the card is automatically ejected in mode 81
.TP
\fB-n URI\fR
Program NFC NDEF URI
.TP
\fB-t text\fR
Program NFC NDEF text
.RE
.TP
\fBYubiKey 3.0 and above\fR
.RS
.TP
\fB-S\fI0605...\fR
set the scanmap to be used with the YubiKey NEO.  It must be 45 unique
bytes as 90 characters.  Leave argument empty to reset to the YubiKey's default.
The scanmap must be sent in the order:
.br
\fIcbdefghijklnrtuvCBDEFGHIJKLNRTUV0123456789!\\t\\r\fR.
.br
the default scanmap in the YubiKey is:
.br
\fI06050708090a0b0c0d0e0f111517181986858788898a8b8c8d8e8f
.br
9195979899271e1f202122232425269e2b28\fR
.br
an example for simplified us dvorak would be:
.br
\fI0c110b071c180d0a0619130f120e09378c918b879c988d8a869993
.br
8f928e89b7271e1f202122232425269e2b28\fR
.br
or for a french azerty keyboard (digits are shifted):
.br
\fI06050708090a0b0c0d0e0f111517181986858788898a8b8c8d8e8f
.br
9195979899a79e9fa0a1a2a3a4a5a6382b28\fR
.br
and a turkish example (has a dotless i instead of usual i):
.br
\fI06050708090a0b340d0e0f111517181986858788898a8b8c8d8e8f
.br
9195979899271e1f202122232425269e2b28\fR
.br
Note that you must remove any whitespace present in these examples
before using the values.
.RE
.TP
\fBYubiKey 2.3 and above\fR
.RS
.TP
\fB-u\fR
Update existing configuration, rather than overwriting. Only possible if the slot is configured as updatable.
.TP
\fB-x\fR
Swap configuration slot 1 and 2 inside the YubiKey. Only possible if both slots are configured as updatable.
.RE
.SH Ticket flags
.TP
[\-]\fBtab-first\fR
Send a tab character as the first character.  This is usually used to move
to the next input field.
.TP
[\-]\fBappend-tab1\fR
Send a tab character between the fixed part and the one-time password
part. This is useful if you have the fixed portion equal to the user
name and two input fields that you navigate between using tab.
.TP
[\-]\fBappend-tab2\fR
Send a tab character as the last character.
.TP
[\-]\fBappend-delay1\fR
Add a half-second delay before sending the one-time password part.
This option is only valid for firmware 1.x and 2.x.
.TP
[\-]\fBappend-delay2\fR
Add a half-second delay after sending the one-time password part.
This option is only valid for firmware 1.x and 2.x.
.TP
[\-]\fBappend-cr\fR
Send a carriage return after sending the one-time password part.
.TP
\fBYubiKey 2.0 firmware and above\fR
.TP
[\-]\fBprotect-cfg2\fR
When written to configuration 1, block later updates to configuration
2.  When written to configuration 2, prevent configuration 1 from
having the lock bit set.
.TP
\fBYubiKey 2.1 firmware and above\fR
.TP
[\-]\fBoath-hotp\fR
Set OATH-HOTP mode rather than YubiKey mode.  In this mode, the token
functions according to the OATH-HOTP standard.
.TP
\fBYubiKey 2.2 firmware and above\fR
.TP
[\-]\fBchal-resp\fR
Set challenge-response mode.
.SH Configuration flags
[\-]\fBsend-ref\fR
Send a reference string of all 16 modhex characters before the fixed
part.  This can not be combined with the \fB-ostrong-pw2\fR flag.
.TP
[\-]\fBpacing-10ms\fR
Add a 10ms delay between key presses.
.TP
[\-]\fBpacing-20ms\fR
Add a 20ms delay between key presses.
.TP
[\-]\fBstatic-ticket\fR
Output a fixed string rather than a one-time password.  The password
is still based on the AES key and should be hard to guess and
impossible to remember.
.TP
\fBYubiKey 1.x firmware only\fR
.TP
[\-]\fBticket-first\fR
Send the one-time password rather than the fixed part first.
.TP
[\-]\fBallow-hidtrig\fR
Allow trigger through HID/keyboard by pressing caps-, num or
scroll-lock twice.  Not recommended for security reasons.
.TP
\fBYubiKey 2.0 firmware and above\fR
.TP
[\-]\fBshort-ticket\fR
Limit the length of the static string to max 16 digits.  This flag
only makes sense with the \fB-ostatic-ticket\fR option.  When
\fB-oshort-ticket\fR is used without \fB-ostatic-ticket\fR it will
program the YubiKey in "scan-code mode", in this mode the key sends
the contents of fixed, uid and key as raw keyboard scancodes.  For
example, by using the fixed string \fIh:8b080f0f122c9a12150f079e\fR in
this mode it will send \fIHello World!\fR on a qwerty keyboard.  This
mode sends raw scan codes, so output will differ between keyboard layouts.
.TP
[\-]\fBstrong-pw1\fR
Upper-case the two first letters of the output string.  This is for
compatibility with legacy systems that enforce both uppercase and
lowercase characters in a password and does not add any security.
.TP
[\-]\fBstrong-pw2\fR
Replace the first eight characters of the modhex alphabet with the
numbers 0 to 7.  Like \fB-ostrong-pw1\fR, this is intended to support
legacy systems.
.TP
[\-]\fBman-update\fR
Enable user-initiated update of the static password.  Only makes sense
with the \fB-ostatic-ticket\fR option.  This is only valid for firmware
2.x.
.TP
\fBYubiKey 2.1 firmware and above\fR
.TP
[\-]\fBoath-hotp8\fR
When set, generate an 8-digit HOTP rather than a 6-digit one.
.TP
[\-]\fBoath-fixed-modhex1\fR
When set, the first byte of the fixed part is sent as modhex.
.TP
[\-]\fBoath-fixed-modhex2\fR
When set, the first two bytes of the fixed part is sent as modhex.
.TP
[\-]\fBoath-fixed-modhex\fR
When set, the fixed part is sent as modhex.
.TP
\fBoath-id=m:OOTTUUUUUUUU\fR
Configure OATH token id with a provided value.  See description of
this option under the 2.2 section for details, but note that a YubiKey
2.1 key can't report its serial number and thus a token identifier value
must be specified.
.TP
\fBYubiKey 2.2 firmware and above\fR
.TP
[\-]\fBchal-yubico\fR
Yubico OTP challenge-response mode.
.TP
[\-]\fBchal-hmac\fR
Generate HMAC-SHA1 challenge responses.
.TP
[\-]\fBhmac-lt64\fR
Calculate HMAC on less than 64 bytes input.  Whatever is in the last byte
of the challenge is used as end of input marker (backtracking from end of payload).
.TP
[\-]\fBchal-btn-trig\fR
The YubiKey will wait for the user to press the key (within 15 seconds) before
answering the challenge.
.TP
[\-]\fBserial-btn-visible\fR
The YubiKey will emit its serial number if the button is pressed during power-up.
This option is only valid for the 2.x firmware line.
.TP
[\-]\fBserial-usb-visible\fR
The YubiKey will indicate its serial number in the USB iSerial field.
This option is not available in the 3.0 and 3.1 firmwares.
.TP
[\-]\fBserial-api-visible\fR
The YubiKey will allow its serial number to be read using an API call.
.TP
\fBoath-id[=m:OOTTUUUUUUUU]\fR
Configure OATH token id with a provided value, or if used without a value use the
standard YubiKey token identifier.

The standard OATH token id for a Yubico YubiKey is (modhex) OO=ub, TT=he,
(decimal) UUUUUUUU=serial number.

The reason for the decimal serial number is to make it easy for humans to correlate
the serial number on the back of the YubiKey to an entry in a list of associated
tokens for example.  Other encodings can be accomplished using the appropriate

NEWS
Yubikey-personalize NEWS -- History of user-visible changes.     -*- outline -*-

* Version 1.17.2 (released 2015-09-22)

** Let _yk_write() return an error if yk_wait_for_key_status() fails.

** Fix a mistake in help, fixed is up to 16 bytes, 32 characters.

** Add 4.2 as supported firmware.

* Version 1.17.1 (released 2015-04-01)

** Fixup of 1.17.0

* Version 1.17.0 (released 2015-04-01)

** add yk_get_capabilities() to fetch capabilities.

** add -c to ykinfo to fetch capabilities.

** whitelist firmware 4.1.x

* Version 1.16.4 (released 2015-03-23)

** change the tool to accept autoeject time as a short instead of a byte

* Version 1.16.3 (released 2015-03-10)

** whitelist YubiKey version 3.4.x

** only try to set libusb configuration if it's unset on the device

* Version 1.16.2 (released 2014-11-28)

** ykinfo: fix modhex printout when serial is an odd number of hex digits.

** whitelist yubikey version 4.0.x

** try to open more PIDs and add for udev.

* Version 1.16.1 (released 2014-10-30)

** Fix udev rules so they contain four digits.

** Only try to detach the kernel driver if it's attached.
For libusb-1.0

** Let import config report errors properly.

* Version 1.16.0 (released 2014-09-26)

** Add yk_get_key_vid_pid() to get the vendor and product id of a key.

** Add flags for ykinfo to print vendor and product id.

** Fix a bug in the osx backend where it would return an error opening
a composite device with two hid interfaces.

* Version 1.15.3 (released 2014-09-04)

** Fix URLs for opensource.y.com -> developers.y.com move.

** Whitelist firmware version 3.3 and detect new PIDs.

* Version 1.15.2 (released 2014-07-30)

** Whitelist firmware version 2.5

** Read key when importing configuration.

** Fix formatting error in information about what is written to key.

** Check return codes when doinf NDEF writes.

* Version 1.15.1 (released 2014-03-12)

** Add -6 and -8 flags to ykchalresp to output the response as a truncated
OATH code. Thanks to joernchen of Phenoelit <joernchen@phenoelit.de>

** Add -t to ykchalresp to use current time / 30 as challenge in TOTP
fashion. Thanks to joernchen of Phenoelit <joernchen@phenoelit.de>

** Fix for a crash in the libusb-0.1 backend.
Thanks to Antti Keränen <detegr@gmail.com>

* Version 1.15.0 (released 2014-01-10)

** add new Neo version 3.2 as supported.

** let ykp_AES_key_from_password() return error if it can't get a random
salt.

** let -a without option mean ask for raw key on STDIN.

** let no -a given mean generate a random key

** add ykp_AES_key_from_raw() and ykp_HMAC_key_from_raw() that sets the
key from a byte array.

** remove -osalt= and possibility to generate key from password.

** add YK_ENODATA that's returned when no data is returned from
the device and no other underlying error is set.

* Version 1.14.2 (released 2013-12-30)

** let -m take challenge-response timeout and autoeject timeout
supplied with : beteen, so mode:cr-timeout:autoeject-timeout

** make sure to always wait 1 millisecond between writes to the key
newer platforms triggered a race where writes to the key could
happen in a very short (about 0.4 milliseconds) timespan and would
not be processed.

** minor fixes for the printing format.

* Version 1.14.1 (released 2013-09-16)

** Fix breakage with latest automake.

* Version 1.14.0 (released 2013-07-04)

** Fixups of import/export.
Add targetConfig to show in which slot a configuration is intended.
Possible memory leaks on error conditions.

** Add -d switch to ykpersonalize for dry-run.

** Add ykp_clear_config() for clearing configuration flags.

** Add getter functions for all configuration flags.

** Add -V to all tools to output version.

** Add ykp_get_acccode_type() and ykp_set_acccode_type()
Only to do with export, showing where the access code came from
in the ycfg.

** Add -1 and -2 options to ykinfo to show programming state.

* Version 1.13.0 (released 2013-04-22)

** Added import/export of ycfg-json format.
Invoke with -fjson to -s or -i
Add exported functions ykp_export_config() and ykp_import_config()

** Fixup output of flags when using ykp_write_config()

** Add binary builds for mac.

** Minor cleanups noticed during debian packaging.

* Version 1.12.0 (released 2013-03-14)

** Recognize firmwares 2.4 and 3.1.

** Add support for setting the new extflag LED_INV
When set the behaviour of the led on the YubiKey is inversed.

* Version 1.11.3 (released 2013-01-09)

** Fixup of broken release.

* Version 1.11.2 (released 2013-01-09)

** Fix a bug where writing a NDEF with unknown prefix ended up writing invalid
data to the YubiKey NEO. Wrote prefix as 0x24 instead of 0x00.

** Don't allow opening a YubiKey if there's more than one present in the system.

** Fix shared linking of ykinfo and ykchalresp.

* Version 1.11.1 (released 2012-12-21)

** Implement ykusb_strerror() on windows.

** Fix a bug where a YubiKey would fail to be recognized if there was
another device from Yubico (vendor id 1050) inserted and looked at
before in the device chain.

** Fix a bug where you could only set 8 bytes of the public id with
the command line tool, now all 16 bytes can be set.

** Documentation updates and fixes.

* Version 1.11.0 (released 2012-12-12)

** Added version symbols and functions.
The header file is "ykpers-version.h" and it contains the following
symbols and functions: YKPERS_VERSION_STRING, YKPERS_VERSION_NUMBER,
YKPERS_VERSION_MAJOR, YKPERS_VERSION_MINOR, YKPERS_VERSION_PATCH,
ykpers_check_version.  Requested by Giampaolo Bellini
<iw2lsi@gmail.com>.

* Version 1.10.0 (released 2012-12-11)

** Support for the new productId of the production Neo.
Has ProducId 0x110, 0x111 or 0x112 depending on mode (see the notes about
-m and device_config).

** Add support for SLOT_NDEF2.
Use SLOT_NDEF to emit slot 1 as NDEF or SLOT_NDEF2 to emit slot 2.
This also adds the function yk_write_ndef2() that takes a slot parameter.

** Add -m flag for ykpersonalize, set usb mode of YubiKey NEO.

ykpersonalize.c
/* -*- mode:C; c-file-style: "bsd" -*- */
/*
 * Copyright (c) 2008-2014 Yubico AB
 * Copyright (c) 2010 Tollef Fog Heen <tfheen@err.no>
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are
 * met:
 *
 *     * Redistributions of source code must retain the above copyright
 *       notice, this list of conditions and the following disclaimer.
 *
 *     * Redistributions in binary form must reproduce the above
 *       copyright notice, this list of conditions and the following
 *       disclaimer in the documentation and/or other materials provided
 *       with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
 * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
 * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
 * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

#include <stdlib.h>
#include <stdio.h>
#include <stddef.h>
#include <string.h>
#include <unistd.h>
#include <errno.h>

#include <ykpers.h>
#include <ykdef.h>

#include "ykpers-args.h"

int main(int argc, char **argv)
{
	FILE *inf = NULL; const char *infname = NULL;
	FILE *outf = NULL; const char *outfname = NULL;
	int data_format = YKP_FORMAT_LEGACY;
	bool verbose = false;
	char keylocation = 0;
	bool use_access_code = false;
	unsigned char access_code[256];
	unsigned char scan_codes[sizeof(SCAN_MAP)];
	YK_KEY *yk = 0;
	YKP_CONFIG *cfg = ykp_alloc();
	YK_STATUS *st = ykds_alloc();
	bool autocommit = false;
	char data[1024];
	bool dry_run = false;

	/* Options */
	char ndef_string[128] = {0};
	char ndef_type = 0;
	unsigned char usb_mode = 0;
	unsigned char cr_timeout = 0;
	unsigned short autoeject_timeout = 0;
	int num_modes_seen = 0;
	bool zap = false;

	/* Assume the worst */
	bool error = true;
	int exit_code = 0;

	ykp_errno = 0;
	yk_errno = 0;

	if (!yk_init()) {
		exit_code = 1;
		goto err;
	}

	if (argc == 2 && strcmp (argv[1], "-h") == 0) {
		fputs(usage, stderr);
		goto err;
	}

	if (!(yk = yk_open_first_key())) {
		exit_code = 1;
		goto err;
	}

	if (!yk_get_status(yk, st)) {
		exit_code = 1;
		goto err;
	}

	printf("Firmware version %d.%d.%d Touch level %d ",
	       ykds_version_major(st),
	       ykds_version_minor(st),
	       ykds_version_build(st),
	       ykds_touch_level(st));
	if (ykds_pgm_seq(st))
		printf("Program sequence %d\n",
		       ykds_pgm_seq(st));
	else
		printf("Unconfigured\n");

	if (!(yk_check_firmware_version2(st))) {
		if (yk_errno == YK_EFIRMWARE) {
			printf("Unsupported firmware revision - some "
			       "features may not be available\n"
			       "Please see \n"
			       "https://developers.yubico.com/yubikey-personalization/doc/Compatibility.html\n"
			       "for more information.\n");
		} else {
			goto err;
		}
	}

	/* Parse all arguments in a testable way */
	if (! args_to_config(argc, argv, cfg, yk,
			     &infname, &outfname,
			     &data_format, &autocommit,
			     st, &verbose, &dry_run,
			     access_code, &use_access_code,
			     &keylocation, &ndef_type, ndef_string,
			     &usb_mode, &zap, scan_codes, &cr_timeout,
			     &autoeject_timeout, &num_modes_seen, &exit_code)) {
		goto err;
	}

	if (verbose && (ykds_version_major(st) > 2 ||
			(ykds_version_major(st) == 2 &&
			 ykds_version_minor(st) >= 2) ||
			(ykds_version_major(st) == 2 && // neo has serial functions
			 ykds_version_minor(st) == 1 &&
			 ykds_version_build(st) >= 4))) {
		unsigned int serial;
		if (! yk_get_serial(yk, 0, 0, &serial)) {
			printf ("Failed to read serial number (serial-api-visible disabled?).\n");

		} else {
			printf ("Serial number : %i\n", serial);
		}
	}

	printf ("\n");

	if (infname) {
		if (strcmp(infname, "-") == 0)
			inf = stdin;
		else
			inf = fopen(infname, "r");
		if (inf == NULL) {
			fprintf(stderr,
				"Couldn't open %s for reading: %s\n",
				infname,
				strerror(errno));
			exit_code = 1;
			goto err;
		}
	}

	if (outfname) {
		if (strcmp(outfname, "-") == 0)
			outf = stdout;
		else
			outf = fopen(outfname, "w");
		if (outf == NULL) {
			fprintf(stderr,
				"Couldn't open %s for writing: %s\n",
				outfname,
				strerror(errno));
			exit(1);
		}
	}

	if (inf) {
		if(!ykp_clear_config(cfg))
			goto err;
		if(!fread(data, 1, 1024, inf))
			goto err;
		if (!ykp_import_config(cfg, data, strlen(data), data_format))
			goto err;
	}
	if (! zap && (ykp_command(cfg) == SLOT_CONFIG || ykp_command(cfg) == SLOT_CONFIG2)) {
		int key_bytes = ykp_get_supported_key_length(cfg);
		char keybuf[42];
		size_t keylen;
		if(keylocation == 2) {
			if(key_bytes == 20) {
				fprintf(stderr, " HMAC key, 20 bytes (40 characters hex) : ");
			} else {
				fprintf(stderr, " AES key, 16 bytes (32 characters hex) : ");
			}
			fflush(stderr);
			if(!fgets(keybuf, sizeof(keybuf), stdin)) {
				printf("error?\n");
				perror ("fgets");
				exit_code = 1;
				goto err;
			}
			keylen = strnlen(keybuf, sizeof(keybuf));
			if(keybuf[keylen - 1] == '\n') {
				keybuf[keylen - 1] = '\0';
			}
			if(key_bytes == 20) {
				if(ykp_HMAC_key_from_hex(cfg, keybuf)) {
					goto err;
				}
			} else {
				if(ykp_AES_key_from_hex(cfg, keybuf)) {
					goto err;
				}
			}
		} else if(keylocation == 0) {
			const char *random_places[] = {
				"/dev/srandom",
				"/dev/urandom",
				"/dev/random",
				0
			};
			const char **random_place;
			size_t read_bytes = 0;

			for (random_place = random_places; *random_place; random_place++) {
				FILE *random_file = fopen(*random_place, "r");
				if (random_file) {
					read_bytes = 0;

					while (read_bytes < key_bytes) {
						size_t n = fread(&keybuf[read_bytes], 1,
								key_bytes - read_bytes, random_file);
						read_bytes += n;
					}

					fclose(random_file);
					break;
				}
			}
			if(read_bytes < key_bytes) {
				ykp_errno = YKP_ENORANDOM;
				goto err;
			}
			if(key_bytes == 20) {
				if(ykp_HMAC_key_from_raw(cfg, keybuf)) {
					goto err;
				}
			} else {
				if(ykp_AES_key_from_raw(cfg, keybuf)) {
					goto err;
				}
			}
		}
	}

	if (outf) {
		if(!(ykp_export_config(cfg, data, 1024, data_format))) {
			goto err;
		}
		if(!(fwrite(data, 1, strlen(data), outf))) {
			goto err;
		}
	} else {
		char commitbuf[256]; size_t commitlen;

		if (ykp_command(cfg) == SLOT_SWAP) {
			fprintf(stderr, "Configuration in slot 1 and 2 will be swapped\n");
		} else if(ykp_command(cfg) == SLOT_NDEF || ykp_command(cfg) == SLOT_NDEF2) {
			fprintf(stderr, "New NDEF will be written as:\n%s\n", ndef_string);
		} else if(ykp_command(cfg) == SLOT_DEVICE_CONFIG) {
			fprintf(stderr, "The USB mode will be set to: 0x%x\n", usb_mode);
			if(num_modes_seen > 1) {
				fprintf(stderr, "The challenge response timeout will be set to: %d\n", cr_timeout);
				if(num_modes_seen > 2) {
					fprintf(stderr, "The smartcard autoeject timeout will be set to: %d\n", autoeject_timeout);
				}
			}
		} else if(ykp_command(cfg) == SLOT_SCAN_MAP) {
			fprintf(stderr, "A new scanmap will be written.\n");
		} else if(zap) {
			fprintf(stderr, "Configuration in slot %d will be deleted\n", ykp_config_num(cfg));
		} else {
			if (ykp_command(cfg) == SLOT_CONFIG || ykp_command(cfg) == SLOT_CONFIG2) {
				fprintf(stderr, "Configuration data to be written to key configuration %d:\n\n", ykp_config_num(cfg));
			} else {
				fprintf(stderr, "Configuration data to be updated in key configuration %d:\n\n", ykp_command(cfg) == SLOT_UPDATE1 ? 1 : 2);
			}
			ykp_export_config(cfg, data, 1024, YKP_FORMAT_LEGACY);
			fwrite(data, 1, strlen(data), stderr);
		}
		fprintf(stderr, "\nCommit? (y/n) [n]: ");
		if (autocommit) {
			strcpy(commitbuf, "yes");
			puts(commitbuf);
		} else {
			if (!fgets(commitbuf, sizeof(commitbuf), stdin))
			{
				perror ("fgets");
				goto err;
			}
		}
		commitlen = strlen(commitbuf);
		if (commitbuf[commitlen - 1] == '\n')
			commitbuf[commitlen - 1] = '\0';
		if (strcmp(commitbuf, "y") == 0
		    || strcmp(commitbuf, "yes") == 0) {
			exit_code = 2;

			if (verbose)
				printf("Attempting to write configuration to the yubikey...");
			if (dry_run) {
				printf("Not writing anything to key due to dry_run requested.\n");
			}
			else if(ykp_command(cfg) == SLOT_NDEF || ykp_command(cfg) == SLOT_NDEF2) {
				YK_NDEF *ndef = ykp_alloc_ndef();
				int confnum = 1;
				int res = 0;
				if(ndef_type == 'U') {
					res = ykp_construct_ndef_uri(ndef, ndef_string);
				} else if(ndef_type == 'T') {
					res = ykp_construct_ndef_text(ndef, ndef_string, "en", false);
				}
				if(!res) {
					if(verbose) {
						printf(" failure to construct ndef\n");
					}
					goto err;
				}
				if(use_access_code) {
					if(!ykp_set_ndef_access_code(ndef, access_code)) {
						if(verbose) {
							printf(" failure to set ndef accesscode\n");
						}
						goto err;
					}
				}
				if(ykp_command(cfg) == SLOT_NDEF2) {
					confnum = 2;
				}
				if (!yk_write_ndef2(yk, ndef, confnum)) {
					if (verbose)
						printf(" failure to write ndef\n");
					goto err;
				}
				ykp_free_ndef(ndef);
			} else if(ykp_command(cfg) == SLOT_DEVICE_CONFIG) {
				YK_DEVICE_CONFIG *device_config = ykp_alloc_device_config();
				ykp_set_device_mode(device_config, usb_mode);
				if(num_modes_seen > 1) {
					ykp_set_device_chalresp_timeout(device_config, cr_timeout);
					if(num_modes_seen > 2) {
						ykp_set_device_autoeject_time(device_config, autoeject_timeout);
					}
				}

				if((usb_mode & 0xf) == MODE_CCID || (usb_mode & 0xf) == MODE_U2F ||
						(usb_mode & 0xf) == MODE_U2F_CCID) {
					fprintf(stderr, "WARNING: This tool will not be able to switch back from the selected mode, really commit? (y/n) [n]: ");
					if (autocommit) {
						strcpy(commitbuf, "yes");
						puts(commitbuf);
					} else {
						if (!fgets(commitbuf, sizeof(commitbuf), stdin))
						{
							perror ("fgets");
							goto err;
						}
					}
					commitlen = strlen(commitbuf);
					if (commitbuf[commitlen - 1] == '\n')
						commitbuf[commitlen - 1] = '\0';
					if (strcmp(commitbuf, "y") != 0
							|| strcmp(commitbuf, "yes") != 0) {
						exit_code = 0;
						error = false;
						goto err;
					}
				}

				if(!yk_write_device_config(yk, device_config)) {
					if(verbose)
						printf(" failure\n");
					goto err;
				}
				ykp_free_device_config(device_config);


			} else if(ykp_command(cfg) == SLOT_SCAN_MAP) {
				if(!yk_write_scan_map(yk, scan_codes)) {
					if(verbose)
						printf(" failure\n");
					goto err;
				}
			} else {
				YK_CONFIG *ycfg = NULL;
				/* if we're deleting a slot we send the configuration as NULL */
				if (!zap) {
					ycfg = ykp_core_config(cfg);
				}
				if (!yk_write_command(yk,
							ycfg, ykp_command(cfg),
							use_access_code ? access_code : NULL)) {
					if (verbose)
						printf(" failure\n");
					goto err;
				}
			}

			if (verbose && !dry_run)
				printf(" success\n");
		}
	}

	exit_code = 0;
	error = false;

err:
	if (error) {
		report_yk_error();
	}

	if (st)
		free(st);
	if (inf)
		fclose(inf);
	if (outf)
		fclose(outf);

	if (yk && !yk_close_key(yk)) {
		report_yk_error();
		exit_code = 2;
	}

	if (!yk_release()) {
		report_yk_error();
		exit_code = 2;
	}

	if (cfg)
		ykp_free_config(cfg);

	exit(exit_code);
}
